R/predict.R

```diff
@@ -0,0 +1,152 @@
+#'
+#' # Set seed
+#' set.seed(42)
+#'
+#' # Simulate data
+#' n <- 10
+#' dat <- data.frame(y = rnorm(n), x = rnorm(n))
+#'
+#' # Fit lm
+#' lm_out <- lm_robust(y ~ x, data = dat)
+#' # Get predicted fits
+#' fits <- predict(lm_out, newdata = dat)
+#' # With standard errors and confidence intervals
+#' fits <- predict(lm_out, newdata = dat, se.fit = TRUE, interval = "confidence")
+#'
+#' # Use new data as well
+#' new_dat <- data.frame(x = runif(n, 5, 8))
+#' predict(lm_out, newdata = new_dat)
+#'
+#' # You can also supply custom variance weights for prediction intervals
+#' new_dat$w <- runif(n)
+#' predict(lm_out, newdata = new_dat, weights = w, interval = "prediction")
+#'
+#' @export
+predict.lm_robust <- function(
+  object,
+  newdata,
+  se.fit = FALSE,
+  interval = c("none", "confidence", "prediction"),
+  alpha = 0.05,
+  na.action = na.pass,
+  pred.var = NULL,
+  weights,
+  ...) {
+
+  # Get model matrix
+  rhs_terms <- delete.response(object$terms)
+  mf <- model.frame(rhs_terms, newdata, na.action = na.action)
+
+  # Check class of columns in newdata match those in model fit
+  if(!is.null(cl <- attr(rhs_terms, "dataClasses"))) .checkMFClasses(cl, mf)
+
+  X <- model.matrix(rhs_terms, mf, contrasts.arg = object$contrasts)
+
+  # lm_lin scaling
+  if (!is.null(object$scaled_center)) {
+    demeaned_covars <-
+      scale(
+        X[
+          ,
+          names(object$scaled_center),
+          drop = F
+          ],
+        center = object$scaled_center,
+        scale = FALSE
+      )
+
+
+    # Interacted with treatment
+    treat_name <- attr(object$terms, "term.labels")[1]
+    interacted_covars <- X[, treat_name] * demeaned_covars
+
+    X <- cbind(X[, attr(X, "assign") <= 1, drop = F],
+               demeaned_covars,
+               interacted_covars)
+  }
+
+  # Get predicted values
+  predictor <- drop(X %*% object$est)
+
+  df_resid <- object$n - object$rank
+  interval <- match.arg(interval)
+
+  if (se.fit || interval != "none") {
+
+    ret <- list()
+
+    var_fit <-
+      apply(X, 1, function(x) tcrossprod(crossprod(x, object$vcov), x))
+
+    if (interval != "none") {
+
+      tval <- qt(alpha/2, df_resid)
+
+      if (interval == "prediction") {
+
+        if (missing(weights) &&
+            object$weighted &&
+            is.null(pred.var)) {
+          warning("Assuming constant prediction variance even though model fit is weighted\\n")
+        }
+
+        # Get weights
+        if (missing(weights)) {
+          weights <- 1
+        } else {
+          weights <- eval(substitute(weights), newdata)
+        }
+
+        pred_var <-
+          if (is.null(pred.var)) {
+            object$res_var / weights
+          } else {
+            pred.var
+          }
+
+        hwid <- tval * sqrt(var_fit + pred_var)
+      } else if (interval == "confidence") {
+        hwid <- tval * sqrt(var_fit)
+      }
+
+      predictor <- cbind(predictor,
+                         predictor + hwid %o% c(1, -1))
```

Comment:
just make three columns explicitly, ```data.frame(fit=predictor, lwr=predictor - hwid, upr=predictor + hwid)```

Reply:
or a matrix if that's your intent.

Reply:
Leaving as a matrix to match `predict.lm`.